src/utils.jl

```diff
@@ -0,0 +1,98 @@
+using OrderedCollections, SparseArrays
+
+"""
+$(TYPEDSIGNATURES)
+
+Return the differential equations and other non-differential equations.
+
+For differential equations, this function assume the derivative is the only single term at
+the left hand side, which is typically the result of `ModelingToolkit.structural_simplify`.
+
+Equations from subsystems are not included.
+"""
+function get_deqs(sys::ODESystem)::Tuple{Vector{Equation}, Vector{Equation}}
+    eqs = ModelingToolkit.get_eqs(sys)
+    deqs = Equation[]
+    others = Equation[]
+    for eq in eqs
+        if eq.lhs isa Term && operation(eq.lhs) isa Differential
+            push!(deqs, eq)
+        else
+            push!(others, eq)
+        end
+    end
+    deqs, others
+end
+
+"""
+$(SIGNATURES)
+
+Given a vector of expressions `exprs` and variables `vars`,
+returns a sparse coefficient matrix `A`, constant terms `c` and nonlinear terms `n`,
+such that `exprs = A * vars + c + n`,
+where the constant terms do not contain any variables in `vars`.
+
+Variables in `vars` must be unique.
+"""
+function linear_terms(exprs::AbstractVector, vars)
```

Comment:
@YingboMa @shashi should this bubble up to being something in MTK itself?

Reply:
Yes, definitely in Symbolics. There's a function called `linear_expansion`.

Reply:
The helper function here also extracts constant terms to facilitate the DEIM algorithm, which is slightly different from [`Symbolics.linear_expansion`](https://github.com/JuliaSymbolics/Symbolics.jl/blob/61bf8a9015749d7869b5e26f08210933c55d2c98/src/linear_algebra.jl#L198).

Reply:
This function extracts 1. coefficients of linear terms, 2. constant terms and 3. remaining nonlinear terms to facilitate the DEIM algorithm, which is different from [`Symbolics.linear_expansion`](https://github.com/JuliaSymbolics/Symbolics.jl/blob/61bf8a9015749d7869b5e26f08210933c55d2c98/src/linear_algebra.jl#L198).
For example,
```julia
using Symbolics, ModelOrderReduction
@variables x y
exprs = [2x - 4y + 5 + (x + 1)^3 + sin(x)
         8x + 7]
l, c, n = ModelOrderReduction.linear_terms(exprs, [x, y])
```
gives
```julia
julia> l
2×2 SparseArrays.SparseMatrixCSC{Float64, Int64} with 3 stored entries:
 2.0  -4.0
 8.0    ⋅ 

julia> c
2-element Vector{Num}:
 5
 7

julia> n
2-element Vector{Num}:
 (1 + x)^3 + sin(x)
                  0
```

However, [`Symbolics.linear_expansion`](https://github.com/JuliaSymbolics/Symbolics.jl/blob/61bf8a9015749d7869b5e26f08210933c55d2c98/src/linear_algebra.jl#L198) doesn't return what is needed in DEIM:
```julia
using Symbolics
@variables x y
exprs = [2x - 4y + 5 + (x + 1)^3 + sin(x)
         8x + 7]
Symbolics.linear_expansion(exprs, [x, y])
```
result: `(Num[#undef #undef; #undef #undef], Num[#undef, #undef], false)`

Reply:
Open an issue on Symbolics.jl about potentially changing `linear_expansion` to what's needed/used here.

Reply:
issue https://github.com/JuliaSymbolics/Symbolics.jl/issues/761